flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java

```diff
@@ -323,6 +355,15 @@ public void validateRunsInMainThread() {
 		assert MainThreadValidatorUtil.isRunningInExpectedThread(currentMainThread.get());
 	}
 
+	/**
+	 * Rpc endpoint's state.
+	 */
+	public enum State {
+		CREATED,
+		STARTED,
+		STOPPING
```

Comment:
What's the reasoning for `STOPPING`, set at the beginning of the stop operation, opposed to `STOPPED`, set after the stop operation?

Reply:
Good question. The start/stop happen asynchronously.
From what I understand, the main thread is ready to operate at the end of the `doStart`.
But `onStop` can additionally trigger some extra asynchronous actions. Only after they are done the endpoint is fully stopped until then it continues to operate but it is known that it is being stopped.
Not sure whether we need more states to reflect the lifecycle better. 
@tillrohrmann probably could help with the input here.